utils/lib/grid_mapping.py

```diff
@@ -0,0 +1,179 @@
+#!/usr/bin/env python3
+import sqlite3
+import argparse
+import os
+import itertools
+
+# =============================================================================
+
+
+class GridLocMap(object):
+    """
+    This class preforms forward (physical -> VPR) and backward (VPR -> physical)
+    grid location mapping.
+    """
+
+    def __init__(self, fwd_loc_map, bwd_loc_map):
+        self.fwd_loc_map = fwd_loc_map
+        self.bwd_loc_map = bwd_loc_map
+
+    @staticmethod
+    def generate_one_to_one_map(extent):
+        """
+        Generates a one-to-one map for specified location range.
+        :param extent:
+        :return:
+        """
+
+        xmin, ymin, xmax, ymax = extent
+
+        fwd_loc_map = {}
+        bwd_loc_map = {}
+
+        # Make a one-to-one map
+        for x, y in itertools.product(range(xmin, xmax + 1), range(ymin,
+                                                                   ymax + 1)):
+
+            fwd_loc_map[(x, y)] = [(x, y)]
+            bwd_loc_map[(x, y)] = [(x, y)]
+
+        # Return map object
+        return GridLocMap(fwd_loc_map, bwd_loc_map)
+
+    @staticmethod
+    def generate_shift_map(extent, shift_x, shift_y):
+        """
+        Generates a one-to-one map for specified location range. For debugging
+        purposes.
+        :param extent:
+        :param shift_x:
+        :param shift_y:
+
+        :return:
+        """
+
+        xmin, ymin, xmax, ymax = extent
+
+        fwd_loc_map = {}
+        bwd_loc_map = {}
+
+        # Make a one-to-one map
+        for x, y in itertools.product(range(xmin, xmax + 1), range(ymin,
+                                                                   ymax + 1)):
+
+            phy_loc = (x, y)
+            vpr_loc = (x + shift_x, y + shift_y)
+
+            fwd_loc_map[phy_loc] = [vpr_loc]
+            bwd_loc_map[vpr_loc] = [phy_loc]
+
+        # Return map object
+        return GridLocMap(fwd_loc_map, bwd_loc_map)
+
+    @staticmethod
+    def load_from_database(conn):
+        """
+        Loads grid location mapping from a SQL database. Returns a GridLocMap
+        object.
+        :param conn:
```

Comment:
Please follow the following style guide: https://github.com/google/styleguide/blob/gh-pages/pyguide.md#383-functions-and-methods

Reply:
Done.